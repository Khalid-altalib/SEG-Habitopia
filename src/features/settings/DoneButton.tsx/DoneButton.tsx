import { NavigationProp, useNavigation } from "@react-navigation/native";
import { Button, View } from "native-base";
import React from "react";
import { RootParams } from "../../../../types";
import { useAppDispatch, useAppSelector } from "../../../app/hooks";
import { setSettings } from "../settingsSlice";

type Props = {
  disabled?: boolean;
  getValues: (valueName: string) => string;
  valueName: string;
};

const DoneButton = (props: Props) => {
  const navigation = useNavigation<NavigationProp<RootParams>>();
  const { getValues, disabled, valueName } = props;

  const { loading } = useAppSelector((state) => state.settings.setSettings);

  const dispatch = useAppDispatch();

  const handlePress = async () => {
    await dispatch(setSettings({ [valueName]: getValues("formValue") }));
    navigation.goBack();
  };

  return (
    <View>
      <Button
        onPress={handlePress}
        mt={8}
        isDisabled={disabled}
        isLoading={loading}
        testID="button"
      >
        Done
      </Button>
    </View>
  );
};

export default DoneButton;
