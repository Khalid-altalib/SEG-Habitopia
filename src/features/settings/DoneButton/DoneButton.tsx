import Button from "@components/Button";
import StatusContainer from "@components/StatusContainer/StatusContainer";
import { NavigationProp, useNavigation } from "@react-navigation/native";
import { View } from "native-base";
import React from "react";
import { ButtonType, RootParams } from "../../../../types";
import { useDispatch, useSelector } from "../../../app/hooks";
import { setSettings } from "../settingsSlice";

type Props = {
  getValues: (valueName: string) => string;
  valueName: string;
  oldPassword?: string;
};

const DoneButton = (props: Props) => {
  const navigation = useNavigation<NavigationProp<RootParams>>();
  const { getValues, valueName, oldPassword } = props;

  const { loading } = useSelector((state) => state.settings.setSettings);

  const dispatch = useDispatch();

  const handlePress = async () => {
    await dispatch(
      setSettings({
        [valueName]: getValues("formValue"),
        oldPassword: oldPassword,
      })
    );
    if (navigation.canGoBack()) {
      navigation.goBack();
    }
  };

  return (
    <Button
      type={ButtonType.Primary}
      onPress={handlePress}
      isFullWidth
      isLoading={loading}
      style={{ marginTop: 25 }}
    >
      Done
    </Button>
  );
};

export default DoneButton;
