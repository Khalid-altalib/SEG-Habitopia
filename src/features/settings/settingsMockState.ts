import { SettingsState } from "./settingsSlice";

/**
 * The mock state for the settings redux state
 */

const settingsMockState: SettingsState = {
  settings: {
    email: "test@test.com",
    password: "password",
    notifications: true,
    name: "Test User",
    biography: "Hello, I'm Test User.",
  },
  fetchSettings: {
    loading: false,
    error: "",
  },
  setSettings: {
    loading: false,
    error: "",
  },
};

export default settingsMockState;
