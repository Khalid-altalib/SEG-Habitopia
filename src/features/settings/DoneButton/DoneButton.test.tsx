import { fireEvent, render, waitFor } from "@testing-library/react-native";
import configureStore, { MockStore } from "redux-mock-store";
import thunk from "redux-thunk";
import DoneButton from "./DoneButton";
import settingsMockState from "../settingsMockState";
import TestingWrapperNavigation from "@app/testingWrapperWithNavigation";

describe("DoneButton", () => {
  const mockState = {
    settings: settingsMockState,
    auth: { user: { authToken: "token", userId: "123" } },
  };
  const mockStore = configureStore([thunk])(mockState);

  const getValuesMock = jest.fn();
  const defaultProps = {
    disabled: false,
    getValues: getValuesMock,
    valueName: "name",
  };

  it("renders correctly", async () => {
    const tree = await render(
      <TestingWrapperNavigation store={mockStore}>
        <DoneButton {...defaultProps} />
      </TestingWrapperNavigation>
    );

    expect(tree).toMatchSnapshot();
  });

  it("finds the done button", async () => {
    const tree = await render(
      <TestingWrapperNavigation store={mockStore}>
        <DoneButton {...defaultProps} />
      </TestingWrapperNavigation>
    );

    const button = tree.getByTestId("button");

    expect(button).toBeDefined();
  });

  it("dispatches setSettings and updates settings when pressed", async () => {
    const store = mockStore;
    const valueName = "name";
    const getValues = jest.fn().mockReturnValue("Jane Doe");

    const wrapper = await render(
      <TestingWrapperNavigation store={mockStore}>
        <DoneButton valueName={valueName} getValues={getValues} />
      </TestingWrapperNavigation>
    );

    const button = wrapper.getByTestId("button");

    await fireEvent.press(button);

    await waitFor(() => {
      expect(getValues).toHaveBeenCalledTimes(1);
      expect(getValues).toHaveBeenCalledWith("formValue");
      const actions = store.getActions();
      // console.log(actions);
      // expect(actions[1].type).toEqual("settings/set/fulfilled");
      // expect(actions[1].payload).toEqual({ name: "Jane Doe" });
      // expect(store.getState().settings.settings.name).toEqual("Jane Doe");
    });
  });
});
