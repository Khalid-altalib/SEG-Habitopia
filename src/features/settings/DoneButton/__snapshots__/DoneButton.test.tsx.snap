// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoneButton renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenStack
      style={
        {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        collapsable={false}
        forwardedRef={[Function]}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        nativeBackButtonDismissalEnabled={false}
        onAppear={[Function]}
        onDisappear={[Function]}
        onDismissed={[Function]}
        onHeaderBackButtonClicked={[Function]}
        onNativeDismissCancelled={[Function]}
        onTransitionProgress={[Function]}
        onWillDisappear={[Function]}
        replaceAnimation="push"
        stackPresentation="push"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        swipeDirection="horizontal"
      >
        <RNSScreenStackHeaderConfig
          backButtonInCustomView={false}
          backgroundColor="rgb(255, 255, 255)"
          color="rgb(0, 122, 255)"
          direction="ltr"
          disableBackButtonMenu={false}
          hidden={false}
          hideBackButton={false}
          largeTitleHideShadow={false}
          title="TestScreen"
          titleColor="rgb(28, 28, 30)"
          topInsetEnabled={false}
          translucent={false}
        />
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            {
              "flex": 1,
              "flexDirection": "column-reverse",
            }
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                },
                {
                  "backgroundColor": "rgb(242, 242, 242)",
                },
                undefined,
              ]
            }
          >
            <View
              dataSet={{}}
              style={{}}
            >
              <View
                accessibilityRole="button"
                accessible={true}
                collapsable={false}
                colorScheme="primary"
                dataSet={{}}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#0891b2",
                    "borderRadius": 4,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginTop": 32,
                    "paddingBottom": 10,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "paddingTop": 10,
                  }
                }
                testID="button"
              >
                <View
                  dataSet={{}}
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                  test={true}
                >
                  <View
                    dataSet={{}}
                    style={{}}
                  >
                    <Text
                      dataSet={{}}
                      style={
                        {
                          "backgroundColor": undefined,
                          "color": "#fafafa",
                          "fontFamily": undefined,
                          "fontSize": 14,
                          "fontStyle": "normal",
                          "fontWeight": "400",
                          "letterSpacing": 0,
                          "lineHeight": 21,
                          "textDecorationLine": undefined,
                        }
                      }
                    >
                      Done
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenStack>
  </View>
</RNCSafeAreaProvider>
`;
